api/grpcserver/node_service.go

```diff
@@ -108,9 +110,9 @@ func (s NodeService) getLayers() (curLayer, latestLayer, verifiedLayer uint32) {
 }
 
 // SyncStart requests that the node start syncing the mesh (if it isn't already syncing).
-func (s NodeService) SyncStart(ctx context.Context, _ *pb.SyncStartRequest) (*pb.SyncStartResponse, error) {
+func (s NodeService) SyncStart(context.Context, *pb.SyncStartRequest) (*pb.SyncStartResponse, error) {
 	log.Info("GRPC NodeService.SyncStart")
-	s.syncer.Start(ctx)
+	s.syncer.Start(s.appCtx)
```

Comment:
I'd prefer to use a context per request, not one for the whole lifetime of the component. Context is supposed to be scoped and relatively short lived, see also the official documentation for context: https://pkg.go.dev/context

Instead of moving towards passing contexts to our services, we should move towards having a context in every method that might be canceled (automatically after a timeout or manually by the caller) instead of using contexts to shut down a service.

Reply:
I am aware, it makes sense in an HTTP server and alike but not exactly for an app with long-running services started on demand. `Syncer::Start()` expects to be passed a context, which cancellation means "stop syncing". Therefore it is invalid to pass it a GRPC call scoped context.

See how `Syncer` is started from _node.go_. 

I feel that we need to figure out how to handle a lifetime of long-lived services throughout the app in a unified way first and then refactor the code. But that's a bigger discussion. @fasmat would you like to lead it?

Reply:
> `Syncer::Start()` expects to be passed a context, which cancellation means "stop syncing". Therefore it is invalid to pass it a GRPC call scoped context.

I can agree with that to some degree, the GRPC `context` should not control when `Syncer` is stopped.

The issue I see is that the `context` passed to `Syncer::Start()` is used to stop the `Syncher` itself instead of "just" canceling the Start if it doesn't finish within some deadline or due to the GRPC caller canceling the request.

In most cases where I have used `context` before I didn't need to hold on to it after the function I passed it into returned (besides maybe passing it to the next function that might be cancelled early).

> I am aware, it makes sense in an HTTP server and alike but not exactly for an app with long-running services started on demand.

There are also long-running services in an HTTP server 😉 

Just quoting from the Go Blog (https://go.dev/blog/context-and-structs):

```
The [documentation for context](https://go.dev/pkg/context/) states:

   Contexts should not be stored inside a struct type, but instead passed to each function that needs it.
```

In the case of `Syncer` I would expect to have to call `Stop` or `Close` when I want the syncher to stop instead of having to cancel the `context` (without knowing how it works internally).

How would a `NodeService::SyncStop()` method look like? One can't easily cancel the `appCtx` (or everything shuts down) and creating a sub context in `StartSync` that would then be canceled in `StopSync` seems quite convoluted to me, at least in comparison of just calling `Syncer::Stop()`.

Reply:
Imo using the `context` passed to `Syncer::Start()` to control its lifecycle is incorrect, maybe we should revert this 
part of the PR otherwise I created this issue: https://github.com/spacemeshos/go-spacemesh/issues/3969 to change it later again.

I'll leave it up to you 🙂